Custom Playlist/Artist/Radio Artwork - #406

ellemarks opened the pull request:
Closes #406 

Adds ability to customize playlist art by adding image file with same name to playlists folder. 

Modifies "core/artwork/read_playlist.go", adding additional functions to check a sub-directory named "zPlaylists" in the music folder, and matches the playlist name to the image file name within said folder, before falling back to grid generation/placeholder image if a corresponding image file is not found. Presently, this commit is not accommodating of other setups, and I expect I will need to tweak the commit. It is very easy to change, but I am unsure on what best practice ought to be , though it has been discussed in #406 - perhaps simply a separate cover_arts folder or similar? 

Attached below is an example of custom playlist art in practice using the Feishin client.
![feishin-custom-playlist](https://github.com/user-attachments/assets/c3cd2557-a720-4663-8bca-2c77bd01d2c8)

From my testing, it works relatively well, although I find that I often find I need to restart the server to force update the cover art for normal playlists, though often worked immediately for smart playlists - I am unsure why exactly, but am guessing that a library scan does not trigger a refresh for normal playlists outright, hence requiring the server restart. Can investigate further if need be.

ellemarks committed:
feat(core/artwork/read_playlist.go): Custom Playlist Art - #406

Adds ability customise playlist art by adding image file with matching name to playlists folder

Signed-off-by: ShadowsDieThrice <140675814+ShadowsDieThrice@users.noreply.github.com>

deluan commented:
Hey, I like the idea and I've been thinking of a similar solution, but not just for playlists, also artists (composers, etc..)

I'll review this PR as soon as can, thanks!

ellemarks committed:
Merge branch 'navidrome:master' into Add-support-for-playlist-cover-art/406

ellemarks committed:
feat(core/artwork/read_playlist.go): Custom Playlist Art - #406 

Adds ability to customise playlist art by adding image file with matching name to playlists folder

Signed-off-by: ShadowsDieThrice <140675814+ShadowsDieThrice@users.noreply.github.com>

ellemarks committed:
feat(core/artwork/read_playlist.go): Custom Playlist Art - #406

Adds ability to customise playlist art by adding image file with matching name to playlists folder

Signed-off-by: ShadowsDieThrice <140675814+ShadowsDieThrice@users.noreply.github.com>

ellemarks committed:
feat(core/artwork/read_playlist.go): Custom Playlist Art - #406

Adds ability to customise playlist art by adding image file with matching name to playlists folder

Signed-off-by: ShadowsDieThrice <140675814+ShadowsDieThrice@users.noreply.github.com>

deluan commented:
I'm thinking of adding a config option like `CoverArtFolder`, that would allow this path to be configurable. Another improvement would be to match CoverArt by name or ID.

Also, we could extend this to artists as well. But for that we would need to have a way to avoid clashes between artists and playlists (and maybe others?) images. The obvious choices would be:

1) use a subdir: `/playlists/Ambivalence.jpg`, `/artist/The Beatles.png`, `/radios/rd_beb1a610cd1b4787996a170ce214759d.webp`, ...
2) use a prefix: Images would be prefixed with a type. Ex: `pl_Ambivalence.jpg`, `ar_The Beatles.png`, `rd_beb1a610cd1b4787996a170ce214759d.webp`, ....


What do you think?

deluan committed:
externalize ArtworkFolder configuration, load any supported image format

Signed-off-by: Deluan <deluan@navidrome.org>

deluan commented:
I started applying the changes I mentioned but we have an issue with the cache: If the user changes the image, it won't invalidate the cache, and Navidrome will still be serving the old image.

We may have to revisit this logic, and check for the artwork image _before_ hitting the cache.

let me know if you have any ideas.

deluan committed:
generalize `fromNamedArtwork` so it can be used by other resources

Signed-off-by: Deluan <deluan@navidrome.org>

ShadowsDieThrice commented:
To avoid clashes, I think a subdir makes more sense from a user standpoint. Even with prefixes, a single folder with hundreds of artist images and ten playlist images seems less than ideal in terms of organisation.

I think the current cache issue is the one that I mentioned before, that I was having to restart the server in order to refresh the playlist images. Honestly, I am not currently familiar enough with the codebase of navidrome to know how to fix this sort of issue. I don't know how navidrome's caching works, but would using something like base64 be applicable here, by comparing the hashes of playlist images, in order to trigger a cache reset? That said, most subsonic clients I've used permanently cache images unless forcibly cleared, so I think caching will always be an issue regardless.

I'm happy to get a bit more involved with the codebase and try to implement this if you still think it's feasible.

ellemarks committed:
Merge branch 'master' into Add-support-for-playlist-cover-art/406

deluan commented:
I have some ideas to solve this. Ping me on Discord and we can discuss it further.

ellemarks committed:
Merge branch 'master' into Add-support-for-playlist-cover-art/406

ShadowsDieThrice commented:
Hey, sorry I left this PR hanging, I've had some things going on - anyway, I've merged v0.55.1 into this PR, and wanted to ask if you've made some changes that fixed the caching issue? Because from my testing (using Feishin), upon refreshing the playlists, custom playlist images seem to update almost immediately if changed. 

Given the scope of #406, this PR seems relatively feature complete - the only other thing would be adding a configurable environmental variable for the `CoverArtFolder`. But were you planning on merging the PR in it's current state, or to include artists, composers and so on?

ellemarks committed:
Merge branch 'master' into Add-support-for-playlist-cover-art/406

deluan committed:
Merge branch 'master' into fork/ShadowsDieThrice/Add-support-for-playlist-cover-art/406

deluan committed:
feat(playlist): add support for custom artwork scanning and storage

Signed-off-by: Deluan <deluan@navidrome.org>

deluan committed:
fix(configuration): update default artwork folder path to use MusicFolder

Signed-off-by: Deluan <deluan@navidrome.org>

deluan commented:
Hey @ShadowsDieThrice I did some changes and completed the PR, but I think we need to remove the matching by name: what happens if two users create playlists with the same name?

I think it should only match by ID. Let me know.

ellemarks committed:
Merge branch 'master' into Add-support-for-playlist-cover-art/406

ShadowsDieThrice commented:
Hey, sorry for basically leaving this PR hanging - sometimes life gets in the way. 

Matching by playlist-id would work, but I think is less intuitive from a users perspective, as the playlist ID itself is only retrievable from the rest API, correct? This seems an unintuitive for most users, in having to identify the playlist-id and learn how to access the rest API in order to name their playlist artwork images correctly. Playlists have an owner though, so perhaps the naming scheme could instead be "USER_PLAYLISTNAME" or similar? This seems the most sensible if the playlist images are managed through the Music_Folder/artwork. If the approach was instead to upload playlist images via the Web-UI, and to store them in the DB, then I think matching by playlist-id would make more sense.

I unfortunately don't really have the time to contribute any more to this, so thanks so much for picking up the slack on this PR on top of the Navidrome project itself.
